src/server/snapshot.cc

```diff
@@ -95,246 +94,242 @@ void SliceSnapshot::Join() {
+// To further complicate things, Table::Traverse covers a logical bucket that may comprise of
+// several physical buckets in dash table. For example, items belonging to logical bucket 0
+// can reside in buckets 0,1 and stash buckets 56-59.
+// PrimeTable::Traverse guarantees an atomic traversal of a single logical bucket,
+// it also guarantees 100% coverage of all items that exists when the traversal started
+// and survived until it finished.
+
 // Serializes all the entries with version less than snapshot_version_.
-void SliceSnapshot::SerializeEntriesFb(const Cancellation* cll) {
-  this_fiber::properties<FiberProps>().set_name(
-      absl::StrCat("SliceSnapshot", ProactorBase::GetIndex()));
-  PrimeTable::Cursor cursor;
+void SliceSnapshot::IterateBucketsFb(const Cancellation* cll) {
+  {
+    auto fiber_name = absl::StrCat("SliceSnapshot-", ProactorBase::GetIndex());
+    this_fiber::properties<FiberProps>().set_name(std::move(fiber_name));
+  }
 
+  PrimeTable::Cursor cursor;
   for (DbIndex db_indx = 0; db_indx < db_array_.size(); ++db_indx) {
-    if (cll->IsCancelled())
-      return;
-
-    if (!db_array_[db_indx])
+    if (cll->IsCancelled() || !db_array_[db_indx])
       continue;
 
-    PrimeTable* pt = &db_array_[db_indx]->prime;
-    VLOG(1) << "Start traversing " << pt->size() << " items";
-
     uint64_t last_yield = 0;
-    mu_.lock();
-    savecb_current_db_ = db_indx;
-    mu_.unlock();
+    PrimeTable* pt = &db_array_[db_indx]->prime;
+    {
+      lock_guard lk(mu_);
+      current_db_ = db_indx;
+    }
 
+    VLOG(1) << "Start traversing " << pt->size() << " items";
     do {
       if (cll->IsCancelled())
         return;
 
-      PrimeTable::Cursor next = pt->Traverse(cursor, [this](auto it) { this->SaveCb(move(it)); });
-
+      PrimeTable::Cursor next =
+          pt->Traverse(cursor, bind(&SliceSnapshot::BucketSaveCb, this, placeholders::_1));
       cursor = next;
+      FlushDefaultBuffer(false);
 
-      // Flush if needed.
-      FlushSfile(false);
-      if (serialized_ >= last_yield + 100) {
+      if (stats_.serialized >= last_yield + 100) {
         DVLOG(2) << "Before sleep " << this_fiber::properties<FiberProps>().name();
         this_fiber::yield();
-        last_yield = serialized_;
         DVLOG(2) << "After sleep";
 
+        last_yield = stats_.serialized;
         // flush in case other fibers (writes commands that pushed previous values)
         // filled the buffer.
-        FlushSfile(false);
+        FlushDefaultBuffer(false);
       }
     } while (cursor);
 
     DVLOG(2) << "after loop " << this_fiber::properties<FiberProps>().name();
-    FlushSfile(true);
+    FlushDefaultBuffer(true);
   }  // for (dbindex)
 
   // Wait for SerializePhysicalBucket to finish.
   mu_.lock();
   mu_.unlock();
 
   for (unsigned i = 10; i > 1; i--)
-    CHECK(!rdb_serializer_->SendFullSyncCut(sfile_.get()));
-  FlushSfile(true);
+    CHECK(!default_serializer_->SendFullSyncCut(default_buffer_.get()));
```

Comment:
why do we send SendFullSyncCut 9 times?

Reply:
That's a small issue 😓😅 Sometimes I need to send just this single byte opcode over a socket and it gets stuck in internal buffers (that happens if the data is less than a couple of bytes).

Reply:
Ohh nice bug

Reply:
@dranikpg  I do not believe it :)
a single byte should still arrive. maybe with a delay but it should.

Reply:
maybe at least add a `TODO: investigate why a single byte gets stuck and does not arrive to replica`.
(btw, there are watermark options in TCP that do not send if a buffer is below K but afaik by default it's disabled).

Reply:
I've personally witnessed it travelling all the way from my request down to the socket 🙂 But I didn't yet look at the rdb_loader code in detail to find where it might be stuck there, I just have some assumptions

Reply:
How did you witness? With tcpdump?

Reply:
Just by logging what is processed by the rdb saver and what is sent further to the sink